feat: add ProveClosest method to SMT and SMST

h5law opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 25 Sep 23 17:51 UTC
This pull request includes the following changes:

- The file diff for file `example.go` includes the addition of a TODO comment to research whether the SiblingData is required and remove it if not.

- The file diff for file `smst.go` includes the addition of a new method called ProveClosest in the SMST struct. This method generates a SparseMerkleProof of inclusion for a given key, using the most common bits as the provided path. It returns the closest path, closest value hash, closest sum, the proof, and any error encountered during the process. Additionally, the Prove and Commit methods remain unchanged in this file.

- The file diff for file `types.go` includes the addition of two new methods ProveClosest in the SparseMerkleTree and SparseMerkleSumTree interfaces. These methods compute a Merkle proof of inclusion for a key in the tree that is closest to the provided path. The search for the closest key is based on the longest common prefix and the key with the most common bits as the provided path.

- The file diff for file `hash.go` includes the addition of an import statement for the "hash" package and a new function named "NoPrehashSpec" which returns a new TreeSpec with a nil Value Hasher and a nil Path Hasher. This function should only be used when values are already hashed and a path is used instead of a key during proof verification.

- The file diff for file `smt.go` includes the addition of a new method ProveClosest in the SMST struct, which generates a SparseMerkleProof of inclusion for the first key with the most common bits as the path provided. It also includes updates to the Prove method to return a proof object, the addition of the recursiveLoad method to recursively load a hash, the addition of the split method to split an extension node, and the addition of the expand method to return the inner node that represents the start of the singly linked list that an extension node represents.

- The file diff for file `smst_proofs_test.go` includes the addition of new test functions: TestSMST_ProveClosest, TestSMST_ProveClosestEmptyAndOneNode. These test the closest value proof in a sparse Merkle sum tree and test the closest value proof with an empty tree and a tree with a single node, respectively. Various updates and modifications were also made to the existing code and test cases in this file.

- The file diff for file `utils.go` includes the addition of a new type nilPathHasher with methods for a path and path size, a new function newNilPathHasher, a comment indicating that the GetPathBit function should be unexported, and a new function flipPathBit to flip the bit at a given position in the data. The setPathBit function was also updated to use bitwise XOR for flipping the bit at a given position. No changes were made to the countSetBits function.

- The file diff for file `smt_proofs_test.go` includes updates to the test cases in this file. A new test case TestSMT_ProveClosest was added to validate the closest key and value hash for a given path in the sparse Merkle tree. Another test case TestSMT_ProveClosestEmptyAndOneNode was added to test the ProveClosest function when the tree is empty or contains only one node. Some helper functions and variables were also added to support the new test cases.

Please review these changes in the code.
<!-- reviewpad:summarize:end -->

## Issue

Fixes N/A

Part of the work related to relay mining: paper can be viewed [here](https://arxiv.org/pdf/2305.10672.pdf)

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Adds ProveClosest method to SMT
- Adds ProveClosest method to SMST
- Adds unit tests

## Testing

- [x] **Task specific tests or benchmarks**: `go test ...`
- [x] **New tests or benchmarks**: `go test ...`
- [x] **All tests**: `go test -v`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling

### If Applicable Checklist

- [ ] Update any relevant README(s)
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams
- [x] I have added tests that prove my fix is effective or that my feature works

Olshansk commented:
@h5law I'm going to review this this week. an you merge it with main?

I wanted to know if you still have your visualizer handy or accessible? Could be good for a visual check.

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+
+	node := smt.tree
+	depth := 0
+	for depth < smt.depth() {
```
Request: can you add some comments throughout the business logic? 

You have this in the lower half of the function.

For example, above the `if node == nil {`  could be a good place, amongst others.

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+			return nil, nil, nil, err
+		}
+		if node == nil && depth == 0 {
+			break // DISCUSS_IN_THIS_PR: I dont think this is even possible
```
This can only happen if the root is nil.

Outside of the forloop, you can check `if smt.tree == nil {return err}`

h5law replied:
So this is actually caught in the very first resolve and error catch, if there is nothing to resolve we error out and we don't need this check at all.

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+		if node == nil && depth == 0 {
+			break // DISCUSS_IN_THIS_PR: I dont think this is even possible
+		}
+		parentDepth = depth
```
We only use these inside the if structure, so should we assign them inside the if as well?

h5law replied:
I have updated these to actually work as intended I had missed that they were essentially no-op.

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+				panic("parent not inner node")
+			}
+			if len(siblings) > 0 {
+				siblings = siblings[:len(siblings)-1]
```
What are we trimming here?

h5law replied:
Added a comment, but if we hit this if statement we have hit a placeholder, prior to hitting it we added the sibling. This sibling is where we want to goto and thus should not be included in the side nodes as it would create an incorrect root hash. Ultimately this is just part of the backstepping.

Olshansk replied:
Please add that explanation in the comments. Very valuable information!

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+				node = inner.leftChild
+			}
+		}
+		if _, ok := node.(*leafNode); ok {
```
```suggestion
		// End traversal once we have reached a leaf node
		if _, ok := node.(*leafNode); ok {
```

Olshansk commented on types.go:
```diff
@@ -30,6 +30,9 @@ type SparseMerkleTree interface {
 	// Prove computes a Merkle proof of membership or non-membership of a key.
 	Prove(key []byte) (*SparseMerkleProof, error)
+	// ProveClosest computes a Merkle proof of inclusion for a key in the tree which is
+	// closest to the path provided
```
Add details on what closest means.

E.g. Most overlap in the path (i.e. # of bits) + default to left?

Olshansk commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+	}
+	// Hash siblings from bottom up.
+	var sideNodes [][]byte
+	for i := range siblings {
```
Why not `i, sibling := range siblings`?

h5law replied:
Because we want to hash the sidenodes from last to first in the slice there is no point taking the `sibling` from the range as we will not use it the next line shows how we get it using `i` and idk if there is a direct way to get this from range without sorting.

Olshansk commented on smt_proofs_test.go:
```diff
@@ -143,3 +143,60 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
+	closestPath := sha256.Sum256([]byte("jackfruit"))
+	require.Equal(t, closestPath[:], closestKey)
+
+	// path = sha256.Sum256([]byte("xwordA188wordB110")) change 31st byte
```
Can we do this automatically? Converting xwordA188wordB110 to a byte slice and then `res[30]...`?

Olshansk replied:
@h5law Noticed that some of the comments haven't been addresses so I wanted to confirm if this is ready for re-review?

h5law replied:
I don't want to make this automatic simply because these are hardcoded it doesn't seem necessary. It would also require us checking the changed byte is not the same which just means extra logic.

LMK if you feel strongly the other way and I will do it.

Olshansk commented on smt_proofs_test.go:
```diff
@@ -143,3 +143,60 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
+
+	result = VerifyProof(proof, root, closestKey, closestValueHash, NoPrehashSpec(sha256.New(), false))
+	require.True(t, result)
+	closestPath = sha256.Sum256([]byte("xwordA188wordB110"))
```
NICE!

Olshansk reviewed (changes requested):
@h5law Took a first review, but will also take a second one tomorrow.

I think this is one of the MOST critical pieces of v1, so would really appreciate a few more comments through the implementation.

h5law commented on smt.go:
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
 }
 
+// ProveClosest generates a SparseMerkleProof of inclusion for the key
+// with the most common bits as the path provided
```
I am unsure what you mean here, it cannot bias a child as it follows a preset path, it biases the leaf with the longest common prefix (what we are looking for) but if along the path a `nil` node is hit (a placeholder) then it must have taken a wrong turn (we want a proof of inclusion). For example we took a right child and it was nil, so we go back to the parent and try the left child. Afterwards we continue to follow the path. This can happen either way and I dont see how it could be biased? LMK

h5law commented on options.go:
```diff
@@ -12,3 +14,15 @@ func WithPathHasher(ph PathHasher) Option {
 }
+
+// NoPrehashSpec returns a new TreeSpec that has a nil Value and Path Hasher
+// NOTE: This should only be used when values are already hashed and a path is
```
Added some more context hope this clears it up

h5law commented:
@Olshansk this PR should be good to go now ready for final reviews

Olshansk commented on smst_proofs_test.go:
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(24))
+
+	// testValue4 is the neighbour of testValue2, by flipping the final bit of the
```
Should these say `Value` or `Key`?

Olshansk commented on smst_proofs_test.go:
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+	root = smst.Root()
+
+	// testKey2 is the child of an inner node which is the child of an extension node
+	// the extension node has the path bounds of [3, 7] by flipping these bits we force
```
`path bounds of [3, 7] by flipping these`

1. I think this is missing some punctuation.
2. You mention the bounds [3, 7] here, and then have two separate bit flips: (3, 6) and (3, 7). Can you just clarify that since the extension has these bounds, we can run tests that flip any bits within those bounds?

h5law replied:
1. Unsure what you mean?
2. We can flip any bits in the 3-6 range and still get the result of testKey2 as these are in the path bounds (flipping other bits changes the path in a way that results in more than a backstep but we are forcing the proof here to do what we want)
	- so flipping bits: 3,4,5,6 result in testKey2
	- flipping bit 7 results in testKey4

Olshansk replied:
Let's update the 4 line comment to this:
```
// `testKey2` is the child of an inner node, which is the child of an extension node.
// The extension node has the path bounds of [3, 7]. This means any bits between
// 3-6 can be flipped, and the resulting path would still traverse through the same 
// extension node and lead to testKey2 - the the closest key. However, flipping bit 7
// will lead to testKey4.
```

Olshansk commented on smst_proofs_test.go:
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath := sha256.Sum256([]byte("testKey2"))
+	require.Equal(t, closestPath[:], closestKey)
```
Tbh, this is really fucking cool

Olshansk commented on smt_proofs_test.go:
```diff
@@ -155,3 +155,61 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
 	require.NoError(t, smv.Stop())
 }
+
+func TestSMT_ProveClosest(t *testing.T) {
```
Please make the same updates here as you do in the other test

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+		}
+		// if we hit a nil node we backstep to the parent node and flip the path bit
+		// at the parent depth and select the other child
+		if node == nil {
```
Not sure what you mean here making this `node != nil` makes a huge block within the if. currently its a single panic inside the branch as this should not happen

Olshansk replied:
This is what I had in mind: 
```go
if node != nil {
  depthDelta = 0
} else {
 // large code block}
}
```

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+		}
+		if ext, ok := node.(*extensionNode); ok {
+			length, match := ext.match(workingPath, depth)
+			if match {
```
The thing is we cannot return or continue after this so both need to be in branches of the if/else as the result is we are either at a leaf or inner node afterwards and want to continue to the logic below. Personally I think swapping the shorter branch for the longer branch is purely stylistic here. LMK if you really think it helps to swap the branches around but I dont think it will help readability

Olshansk replied:
This is what I had in mind:
```go
if !match {
   node = ext.expand()
} else {
  // large code block
}
```

I personally think it'll help readability, but it is a NIT.

If you strongly disagree, I want push for it.

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+				node = ext.expand()
+			}
+		}
+		inner := node.(*innerNode)
```
There is only 3 types of nodes if it is neither a leaf or an extension node it must be an inner node

h5law replied:
@Olshansk this actually poses a good question, this is only possible to fail when the tree is empty. We are looking for a proof of inclusion and the updated return here results in a successful proof of inclusion of an empty tree. This could be checked user side (ie, is closestPath a placeholder is so this tree was empty) but I think the new logic here should be good.

Olshansk replied:
Let's add a check. If it errors return an error that says "This can only happen if a tree is empty".

It'll be a good way to catch errors / edge cases we may not be aware of (which should THEORETICALLY not happen)

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+	if node == nil {
+		panic("no leaf node found")
+	}
+	leaf := node.(*leafNode)
```
See comment above this is impossible to fail

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+
+	// Retrieve the closest path and value hash if found
+	if node == nil {
+		panic("no leaf node found")
```
Returning an error here now see a comment above regarding empty trees

h5law commented on smt.go:
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
+	proof = &SparseMerkleProof{
+		SideNodes: sideNodes,
+	}
+	if sib != nil {
```
So this seems to be left over from something I cant see it used anywhere in proof verification. Will leave it in for now and remove in another PR once I have fully gone through the code and checked it wont do any damage

Olshansk replied:
Add a TODO with an exaplanation

h5law commented on smst_proofs_test.go:
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+	require.NoError(t, smst.Update([]byte("testKey2"), []byte("testValue"), 5))
+	require.NoError(t, smst.Update([]byte("testKey3"), []byte("testValue"), 5))
+	require.NoError(t, smst.Update([]byte("testKey4"), []byte("testValue"), 5))
+	// insert testing values that are similar
```
I have removed these tests as the visual test is much better at doing what we are testing. These are simply covering the pre-existing tree traversal logic works which is the same as before

h5law commented on smst_proofs_test.go:
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+}
+
+// ProveClosest test against a visual representation of the tree
+// See: https://i.imgur.com/cPJObIy.png
```
How did you do this?

Olshansk replied:
Upload an image in one of the comments and use that URL

h5law commented on utils.go:
```diff
@@ -26,6 +37,13 @@ func setPathBit(data []byte, position int) {
 }
 
+// flipPathBit flips the bit at an offset from the most significant bit
+func flipPathBit(data []byte, position int) {
```
I wanna keep it as `flipPathBit` to match `setPathBit` and `GetPathBit` (will be unexporting this now we dont need it in v1 anymore)

h5law commented:
@Olshansk I addressed the comments and replied to some of the outstanding ones with reasonings, there are a few PRs that should come out of this one:
1. Look into removing `SiblingData` from proofs
2. Unexport `GetPathBit`
Otherwise I think the comments should clear things up that may have been unclear. I removed the original test cases in favour of the visual ones as what we are forcing is backstepping logic. Otherwise the proof functions the same and this logic is covered elsewhere.

Olshansk commented on smt.go:
```diff
@@ -398,6 +398,9 @@ func (smt *SMT) ProveClosest(path []byte) (
 	var siblings []treeNode
 	var sib treeNode
 	var parent treeNode
+	// depthDelta is used to track the depth increase when traversing down the tree
```
ty!

Olshansk commented on smt.go:
```diff
@@ -451,7 +460,10 @@ func (smt *SMT) ProveClosest(path []byte) (
-		inner := node.(*innerNode)
+		inner, ok := node.(*innerNode)
+		if !ok { // this can only happen for an empty tree
+			break
```
add a print statement!

h5law replied:
@Olshansk Mentioned this below but I think the result of the return values should be used instead of a print. Besides debugging the SMT itself is there a usecase for print statements with the new return values clearly showing the tree was empty vs. 1 node vs N nodes

Olshansk replied:
Fair enough. Resolving

Olshansk commented on smt.go:
```diff
@@ -463,10 +475,14 @@ func (smt *SMT) ProveClosest(path []byte) (
 	// Retrieve the closest path and value hash if found
-	if node == nil {
-		panic("no leaf node found")
+	if node == nil { // tree was empty
```
I think we should print that the tree was empty.

If/when we hit bugs, this is the sort of thing that helps find it.

h5law replied:
I personally would be against including print statements in the library itself only included when doing debugging otherwise they would just bloat out the SMT during normal use. I think having added the new return specific to empty trees allows us to determine the outcome without needing the print statement:
1. Was the `closestPath` returned a placeholder & is the root hash a placeholder & was `closestValueHash` returned nil?
  - Tree 100% empty
2. Otherwise tree is non-empty

Olshansk commented on smt_proofs_test.go:
```diff
@@ -156,6 +156,8 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
 }
 
+// ProveClosest test against a visual representation of the tree
+// See: https://i.imgur.com/ux4fchQ.png
```
Use the github URL?

Olshansk commented:
@h5law Seems like this is out of date with the main branch and might need to be rebased or merged.

Olshansk commented:
> Look into removing SiblingData from proofs
> Unexport GetPathBit

Leave TODOs next to the structure/function with next steps. It's the most lightweight way to keep track of work (no tickets, etc) without losing context if both of us are not around.

Olshansk reviewed (dismissed):
:shipit:
